vowpalwabbit/chained_proxy_iterator.h

```diff
@@ -52,39 +52,53 @@ struct chained_proxy_iterator
   {
     ++_current;
     // TODO: don't rely on audit_end
-    if (_current == (*_outer_current).audit_end() && (_outer_current != _outer_end))
+    if (_current == (*_outer_current).feats.audit_end() && (_outer_current != _outer_end))
     {
       ++_outer_current;
-      _current = (*_outer_current).audit_begin();
+      _current = (*_outer_current).feats.audit_begin();
     }
     return *this;
   }
 
-  // TODO jump full feature groups.
+  // UB if diff < 0
   chained_proxy_iterator& operator+=(difference_type diff)
   {
-    for (size_t i = 0; i < diff; i++) { operator++(); }
-    return *this;
+    assert(diff >= 0);
+    while (true)
+    {
+      auto current_group_distance_to_end = std::distance(_current, (*_outer_current).feats.audit_end());
+      if (diff > current_group_distance_to_end)
```

Comment:
what happens if you try to increment beyond the end of the list? say, `iter += INT_MAX`? It looks like an infinite loop or undefined behavior. Did i miss something?

Reply:
If you += beyond the end of the list it will be undefined behavior, same as with any STL iterator.

> The behavior is undefined if the specified sequence of increments or decrements would require that a non-incrementable iterator (such as the past-the-end iterator) is incremented, or that a non-decrementable iterator (such as the front iterator or the singular iterator) is decremented. 

https://en.cppreference.com/w/cpp/iterator/advance